src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -520,8 +520,12 @@ static ucs_status_t uct_ib_mlx5_devx_reg_atomic_key(uct_ib_md_t *ibmd,
     ucs_status_t status;
     uint8_t mr_id;
 
-    if (!uct_ib_mlx5_devx_use_atomic_ksm(md, memh)) {
-        return uct_ib_mlx5_reg_atomic_key(ibmd, ib_memh);
+    if (!ucs_test_all_flags(md->flags,
+                            UCT_IB_MLX5_MD_FLAG_KSM |
+                            UCT_IB_MLX5_MD_FLAG_INDIRECT_ATOMICS)) {
+        /* We cannot return a direct key, since the requestor may try to use it
+           with non-zero atomic offset */
+        return UCS_ERR_UNSUPPORTED;
     }
 
     status = uct_ib_mlx5_md_get_atomic_mr_id(ibmd, &mr_id);
```

Comment:
Maybe `uct_ib_mlx5_md_get_atomic_mr_id` should set `mr_id` to zero if `uct_ib_mlx5_devx_use_atomic_ksm` return false? Can move `uct_ib_mlx5_md_get_atomic_mr` into **ib_mlx5dv_md.c**

Reply:
unfortunately we can't do it because mr_id is part of flush_rkey (to reduce address size) so if mr_id is zero, flush_rkey must be disabled. IMO it was better to disable relaxed order (break performance) than disable flush_rkey (break correctness)